modin/pandas/dataframe.py

```diff
@@ -1608,9 +1608,26 @@ def query(self, expr, inplace=False, **kwargs):  # noqa: PR01, RT01, D200
         """
         self._update_var_dicts_in_kwargs(expr, kwargs)
         inplace = validate_bool_kwarg(inplace, "inplace")
-        new_query_compiler = pandas.DataFrame.query(
-            self, expr, inplace=False, **kwargs
-        )._query_compiler
+        new_query_compiler = None
+        # HACK: this condition kind of breaks the idea of backend agnostic API as all queries
+        # _should_ work fine for all of the engines using `pandas.DataFrame.query(...)` approach.
+        # However, at this point we know that we can execute simple queries way more efficiently
+        # using the QC's API directly in case of pandas backend. Ideally, we have to make it work
+        # with the 'pandas.query' approach the same as good the direct QC call is. But investigating
+        # and fixing the root cause of the perf difference appears to be much more complicated
+        # than putting this hack here. Hopefully, we'll get rid of it soon:
+        # https://github.com/modin-project/modin/issues/6499
+        if StorageFormat.get() == "Pandas":
+            self._validate_eval_query(expr, **kwargs)
+            try:
+                new_query_compiler = self._query_compiler._rowwise_query(expr, **kwargs)
```

Comment:
We can add `_rowwise_query` to BaseQueryCompiler, which does nothing except throw NotImplementedError.

Then implement `_rowwise_query` for PandasQueryCompiler something like:
```python
def _rowwise_query(self, expr, **kwargs):
     self._validate_eval_query(expr, **kwargs)
     [your implementation]
```

Reply:
Then on DataFrame level code will be much nicer:
```python
try:
  # kind of fast path
  new_query_compiler = self._query_compiler._rowwise_query(expr, **kwargs)
except NotImplementedError:
  [old implementation]
```